cmake/modules/FindGINKGO.cmake

```diff
@@ -19,54 +19,53 @@
 # This module exports
 #
 #   GINKGO_INCLUDE_DIRS
+#   GINKGO_INTERFACE_LINK_FLAGS
 #
 
 SET(GINKGO_DIR "" CACHE PATH "An optional hint to a GINKGO installation")
 SET_IF_EMPTY(GINKGO_DIR "$ENV{GINKGO_DIR}")
 
-DEAL_II_FIND_LIBRARY(GINKGO_LIBRARY
-  NAMES ginkgo
-  HINTS ${GINKGO_DIR}
-  PATH_SUFFIXES
-    lib${LIB_SUFFIX} lib64 lib
-    # This is a hint, isn't it?
-    build/${CMAKE_CXX_PLATFORM_ID}-${CMAKE_SYSTEM_PROCESSOR}/libginkgo
-  )
-DEAL_II_FIND_LIBRARY(GINKGO_REFERENCE_LIBRARY
-  NAMES ginkgo_reference
-  HINTS ${GINKGO_DIR}
-  PATH_SUFFIXES
-    lib${LIB_SUFFIX} lib64 lib
-    # This is a hint, isn't it?
-    build/${CMAKE_CXX_PLATFORM_ID}-${CMAKE_SYSTEM_PROCESSOR}/libginkgo_reference
-  )
-DEAL_II_FIND_LIBRARY(GINKGO_OMP_LIBRARY
-  NAMES ginkgo_omp
-  HINTS ${GINKGO_DIR}
-  PATH_SUFFIXES
-    lib${LIB_SUFFIX} lib64 lib
-    # This is a hint, isn't it?
-    build/${CMAKE_CXX_PLATFORM_ID}-${CMAKE_SYSTEM_PROCESSOR}/libginkgo_omp
-  )
-DEAL_II_FIND_LIBRARY(GINKGO_CUDA_LIBRARY
-  NAMES ginkgo_cuda
-  HINTS ${GINKGO_DIR}
-  PATH_SUFFIXES
-    lib${LIB_SUFFIX} lib64 lib
-    # This is a hint, isn't it?
-    build/${CMAKE_CXX_PLATFORM_ID}-${CMAKE_SYSTEM_PROCESSOR}/libginkgo_cuda
+#
+# Save and restore the ${CMAKE_MODULE_PATH} variable. The Ginkgo project
+# configuration unfortunately overrides the variable which causes
+# subsequent configuration to fail.
+#
+SET(_cmake_module_path ${CMAKE_MODULE_PATH})
+FIND_PACKAGE(Ginkgo
+  HINTS ${GINKGO_DIR} ${Ginkgo_DIR} $ENV{Ginkgo_DIR}
   )
+SET(CMAKE_MODULE_PATH ${_cmake_module_path})
 
-DEAL_II_FIND_PATH(GINKGO_INCLUDE_DIR ginkgo/ginkgo.hpp
-  HINTS ${GINKGO_DIR}
-  PATH_SUFFIXES include
-  )
+#
+# Cosmetic clean up: Let's remove all variables beginning with "GINKGO_"
+# that are actually not used during configuration but show up in
+# detailed.log
+#
+unset(GINKGO_CXX_COMPILER)
+
+#
+# We'd like to have the full library names but the Ginkgo package only
+# exports a list with short names. So check again for every lib and store
+# the full path:
+#
+SET(_libraries "")
+FOREACH(_library ${GINKGO_INTERFACE_LINK_LIBRARIES})
```

Comment:
`GINKGO_INTERFACE_LINK_LIBRARIES` only contains the list of libraries that `libginkgo.so` depends on so `libginkgo_reference.so`, `libginkgo_cuda.so` and `libginkgo_omp.so`, But we also need to link against `libginkgo.so` and hence the undefined reference errors.

Essentially `GinkgoConfig.cmake` in `install_dir/lib/cmake/Ginkgo` looks like:
```
set(GINKGO_INTERFACE_LINK_LIBRARIES "ginkgo_omp;ginkgo_cuda;ginkgo_reference")
set(GINKGO_INTERFACE_LINK_FLAGS "-L<ginkgo_install_dir>/lib -lginkgo -lginkgo_omp -lginkgo_cuda -lginkgo_reference"
```



That was why I was using the `GINKGO_INTERFACE_LINK_FLAGS` to get the proper [list of flags to link against in DEAL_II_PACKAGE_HANDLE](https://github.com/dealii/dealii/pull/11413/commits/cbd7e9bed5528e43326a17f667b3a1da735ca456#diff-23b767ae3a5cb11c2e3894299d5d20b9559bb0455563936d950a8da702c99e0bR34).

Ideally, the best way would be the cmake way of using the `Ginkgo::ginkgo` target and just simply using `target_link_libraries(${dealii_target} Ginkgo::ginkgo)`, but I could not find a place to do this within deal.ii.

Reply:
Using imported targets such as `Ginkgo::ginkgo` directly unfortunately doesn't solve the problem that we have to export the full transitive link interface in a meaningful way. And we have to do that uniformly for all external dependencies.

We are currently solving this issue with a custom `DEAL_II_PACKAGE_HANDLE` macro. The whole deal.II build system was designed with cmake-2.8.8 and is thus a bit rusty.  We tried to transition to imported targets once but had to roll back due to some unforeseen incompatibilities. Currently, I would very much prefer not to use imported targets (such as `Ginkgo::ginkgo` before we have figured out how to correctly integrate those.

That said, the core problem is that the CMake ecosystem has not figured out a canonical standard format for finding and setting up external dependencies: Some projects have a "project configuration", some have a `FindFEATURE.cmake` macro shipped with CMake and most do not provide anything. So we unfortunately, had to settle on the greatest commond divisor - which is to do all the dependency setup by hand (in our own `FindFEATURE.cmake` files).